module/VuFind/src/VuFind/I18n/Initializer.php

```diff
@@ -0,0 +1,289 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Translator
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Sebastian Kehr <kehr@ub.uni-leipzig.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+namespace VuFind\I18n;
+
+use VuFind\Cookie\CookieManager;
+use VuFind\I18n\Translator\Loader\ExtendedIni;
+use Zend\Cache\Storage\StorageInterface;
+use Zend\Config\Config;
+use Zend\Http\PhpEnvironment\Request;
+use Zend\I18n\Translator\Translator as I18nTranslator;
+use Zend\Mvc\I18n\Translator as MvcTranslator;
+use Zend\View\Model\ViewModel;
+
+/**
+ * Handles I18n initialization.
+ *
+ * @category VuFind
+ * @package  Translator
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Sebastian Kehr <kehr@ub.uni-leipzig.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class Initializer
+{
+    /**
+     * Translator.
+     *
+     * @var I18nTranslator
+     */
+    protected $translator;
+
+    /**
+     * Language base directories.
+     *
+     * @var string[]
+     */
+    protected $baseDirs = [];
+
+    /**
+     * File loader.
+     *
+     * @var ExtendedIni
+     */
+    protected $loader;
+
+    /**
+     * Initializer constructor.
+     *
+     * @param Request          $request       Request
+     * @param Config           $config        Configuration
+     * @param CookieManager    $cookies       Cookie manager
+     * @param StorageInterface $cache         Language cache
+     * @param MvcTranslator    $mvcTranslator Translator
+     * @param ViewModel        $viewModel     ViewModel
+     */
+    public function __construct(
+        Request $request,
+        Config $config,
+        CookieManager $cookies,
+        StorageInterface $cache,
+        MvcTranslator $mvcTranslator,
+        ViewModel $viewModel
+    ) {
+        $language = $this->detectLanguage($config, $request, $cookies);
+
+        $this->translator = $mvcTranslator->getTranslator();
+        $this->loader = $this->translator->getPluginManager()
+            ->get(ExtendedIni::class);
+
+        $this->translator->setLocale($language);
+        $this->translator->setCache($cache);
+
+        $this->loader->setFallbacks($this->parseFallbacks($config));
+
+        $rtlLangs = isset($config->LanguageSettings->rtl_langs)
+            ? array_map(
+                'trim', explode(',', $config->LanguageSettings->rtl_langs)
+            ) : [];
+
+        $viewModel->setVariable('userLang', $language);
+        $viewModel->setVariable('allLangs', $config->Languages);
+        $viewModel->setVariable('rtl', in_array($language, $rtlLangs));
+
```

Comment:
In my opinion view model variables should not be set here. Instead the computed values should only be provided as a service (which could then be especially consumed for view model initialization).

Reply:
I agree that this is inelegant -- it's an implementation detail I have never been very happy with -- but if we change it we'll need to be sure we don't break anything in the process.